### ISSUE TITLE
SDRClassifierRegion does not play well with scalar value prediction

### ISSUE THREAD

**Issue by andrewmalta13** [numenta/nupic#3225]
Discussed with Yuwei and we agreed that the SDRClassifierRegion inherited some problems from the CLAClassifierRegion and that it does not work with multistep scalar prediction.

<a href="https://github.com/numenta/nupic/blob/master/src/nupic/regions/CLAClassifierRegion.py#L380"> This </a>  is the line we discussed that was inherited from the CLAClassifier that creates this problem. 

A related <a href="https://github.com/numenta/nupic/issues/3078"> issue  </a> from someone experiencing the same problem I had.

@ywcui1990 what would be your proposal to making this region work both for traditional classification and scalar value prediction?

@rhyolight I am putting the prediction network api example on hold until this is resolved.

**Comment by natoromano** [comment]
Here is my suggested fix:
- `SDRClassifierRegion` should have an _optional_ additional input, `valueIn`
- the `RecordSensor` region should have an _optional_ additional output, `predictedFieldOut`

The `RecordSensor` can use its reference to its encoder to determine a good `bucketIdx` for each value, and use that as `categoryOut`, as well as send the predicted field in `predictedValueOut`. That would perhaps require adding an `.enablePrediction(predictionArgs)` to the `RecordSensor`, much like is done [here](https://github.com/numenta/nupic/blob/master/src/nupic/frameworks/opf/model.py#L155) for the OPF framework. `predictionArgs` can just be the name of the field, or have additional information (number of buckets to use if no encoder is used?).
Of course, a flag would be set by `enablePrediction`, and nothing is sent to `predictedValueOut` if this flag is not set.

As for the `SDRClassifierRegion`, it can do what it currently does if there is no `predictionInput`, i.e. use `int(category)` for both the bucket ID and actual value. If there is a `predictionInput`, it uses `int(category)` for the bucket ID (which, if connected to a sensor, is optimal as it is the same bucket as in the encoder), and `predictionInput` for the actual value.

@ywcui1990 @scottpurdy @andrewmalta13

**Comment by scottpurdy** [comment]
Alternatively, the main implementation could just be for categories and a subclass could additionally do the bucket->scalar conversion. This or @natoromano 's proposal are fine with me. His is simpler.

**Comment by andrewmalta13** [comment]
The next person who picks up this issue should check out some preliminary work I did on this branch: https://github.com/andrewmalta13/nupic/tree/scalarPrediction

In particular I added an example that creates a network using the network API for scalar value prediction that would serve as a good use case to develop towards.  As for the changes to the SDRClassifierRegion and RecordSensor, they are just minor changes, but they implement some of the details specified in @natoromano  comment above.  

Some work left to do would be to figure out the best way to leverage the encoder to give a reasonable bucket-id and send it via the categoryOut output when doing scalar value prediction.  Part of the work is also figuring out the best way to specify the desired predicted field. Also when doing scalar prediction, you don't want to want to be casting the actualValue to an int in the SDRClassifierRegion.
